src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
             }
             else if (certInfo.IsFileCert)
             {
-                var env = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
-                return new X509Certificate2(Path.Combine(env.ContentRootPath, certInfo.Path), certInfo.Password);
+                var environment = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
+                if (certInfo.KeyPath != null)
+                {
+                    X509Certificate2 certificate = null;
+                    if (TryReadPemRSAKey(certInfo, environment, out var rsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate =  publicCertificate.CopyWithPrivateKey(rsaKey);
+                    }
+
+                    if (TryReadPemDSAKey(certInfo, environment, out var dsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate = publicCertificate.CopyWithPrivateKey(dsaKey);
+                    }
+
+                    if (certificate != null)
+                    {
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
+                            return PersistKey(certificate);
+                        }
+
+                        return certificate;
+                    }
```

Comment:
All of this is logically just

```C#
X509Certificate2 certificate = X509Certificate2.CreateFromPemFile(
    Path.Combine(environment.ContentRootPath, certInfo.Path),
    Path.Combine(environment.ContentRootPath, certInfo.KeyPath));
```

using the new net5 API (checked in already, might not be in a released preview)

If that throws you could fall back to `CreateFromEncryptedPemFile(certPath, certInfo.Password, keyPath)`, assuming that's a scenario.

Alternatively, using the API that already exists, you can use cert.PublicKey.Oid.Value to see what algorithm the cert's public key is, which would tell you what kind of private key to expect.  The problem with that approach is that you need to update here any time new algorithms come along.

Reply:
I see that the current tests are using a DER-encoded cert along with a PEM-encoded key. That's not a scenario that we support at the lower level, we expect PFX (existing routines) or PEM+PEM (the new methods).

The approach here with CopyWithPrivateKey will get exceptions if the cert file is a PFX and already had a private key, then you loaded a private key.

I'd recommend only supporting PEM+PEM (the new method).  Extracting the signing cert from an authenticode-signed executable and mating it with a PEM-based key is a weird scenario to allow.

Reply:
> The approach here with CopyWithPrivateKey will get exceptions if the cert file is a PFX and already had a private key, then you loaded a private key.

Wouldn't that fail to load because we wouldn't even be passing a password to it?

> I see that the current tests are using a DER-encoded cert along with a PEM-encoded key. That's not a scenario that we support at the lower level, we expect PFX (existing routines) or PEM+PEM (the new methods).

Will this be a problem as long as we check that the certificate loaded in advance doesn't have an associated private key or are there any other issues we should be aware of.

I want to avoid forcing people to convert their certificate format to do this and this is something that I've seen used in other web frameworks too. (DER Cert + PEM key), so I would like that combo to work out of the box, provided that the cert doesn't already contain a private key.

Reply:
>>     The approach here with CopyWithPrivateKey will get exceptions if the cert file is a PFX and already had a private key, then you loaded a private key.

> Wouldn't that fail to load because we wouldn't even be passing a password to it?

If it had a NULL password or an empty password it will load using the simple ctor.  If it's a PFX with a password other than those two values, then it'd throw an exception for the password being wrong.

> I want to avoid forcing people to convert their certificate format to do this

In my experience (and certainly in the Linux+Let's Encrypt+certbot case) a PEM key is paired with a PEM cert; particularly because they'll use openssl commands, which use PEM output unless you tell it otherwise.

> this is something that I've seen used in other web frameworks too. (DER Cert + PEM key)

Can you give some examples?

> Will this be a problem as long as we check that the certificate loaded in advance doesn't have an associated private key or are there any other issues we should be aware of.

The X509Certificate2 constructor, much to my chagrin, doesn't do simple things.  It does complex things.  The input data can be:

* X.509 public key certificate, binary encoded
* X.509 public key certificate, PEM encoded
* (Not Linux today, bug should be fixed later in 5) A PKCS#7 SignedCms object, binary encoded, where the first signer cert is used
* (Not Linux today, bug should be fixed later in 5) A PKCS#7 SignedCms object, PEM encoded, where the first signer cert is used
* (Windows-only) An authenticode-signed exe, dll, msi, msu, etc, where the first signer cert is used.
* (Windows-only) The SerializedCert format (Windows-proprietary) and probably other things.
* A PKCS#12 PFX, binary encoded

If you really wanted to support both the binary and PEM encoding of a certificate, but restrict the inputs to more sensible things, you could use X509Certificate2.GetCertContentType to ask what it is, and only accept X509.  That still eliminates the case where you load a PFX and have a keyfile specified (and avoids writing the key to disk at all in that error case).

Reply:
> Can you give some examples?

Many dev tools support passing the der cert + the pem keyand handle it transparently. I was doing something similar the other day with the nodejs proxy middleware and they support this transparently. If we can reasonably do it, I don't see why we wouldn't.

I think that with the changes you suggested to how we load the keys it should be enough to support this without problems or keys leaking in error cases.